src/PromTracker.mo

```diff
@@ -0,0 +1,291 @@
+    public func share() : StableData {
+      var res : StableData = null;
+      for (value in Vec.vals(values_)) {
+        switch (value) {
+          case (?v) switch (v.share()) {
+            case (?data) {
+              res := AssocList.replace(res, v.prefix(), Text.equal, ?data).0;
+            };
+            case (_) {};
+          };
+          case (null) {};
+        };
+      };
+      res;
+    };
+
+    /// Patch all values with stable data
+    public func unshare(data : StableData) : () {
+      for (value in Vec.vals(values_)) {
+        switch (value) {
+          case (?v) switch (AssocList.find(data, v.prefix(), Text.equal)) {
+            case (?data) v.unshare(data);
+            case (_) {};
+          };
+          case (_) {};
+        };
+      };
+    };
+
+  };
+
+  class PullValue(prefix_ : Text, pull : () -> Nat) {
+
+    public func value() : Nat = pull();
+
+    public func dump() : [(Text, Nat)] = [(prefix_, pull())];
+
+    public func prefix() : Text = prefix_;
+    public func share() : ?StableDataItem = null;
+    public func unshare(data : StableDataItem) = ();
+  };
+
+  class AccumulatorValue(prefix_ : Text, isStable : Bool) {
+    var value_ = 0;
+
+    public func value() : Nat = value_;
+    public func add(n : Nat) { value_ += n };
+    public func set(n : Nat) { value_ := n };
+
+    public func dump() : [(Text, Nat)] = [(prefix_, value_)];
+
+    public func prefix() : Text = prefix_;
+    public func share() : ?StableDataItem {
+      if (not isStable) return null;
+      ? #counter(value_);
+    };
+    public func unshare(data : StableDataItem) = switch (data) {
+      case (#counter x) value_ := x;
+      case (_) {};
+    };
+  };
+
+  class GaugeValue(prefix_ : Text, watermarkResetIntervalSeconds : Nat, isStable : Bool) {
+
+    class WatermarkTracker<T>(default : T, condition : (old : T, new : T) -> Bool, resetIntervalSeconds : Nat) {
+      var lastWatermarkTimestamp : Time.Time = 0;
+      var val : T = default;
+      public func value() : T = val;
+      public func update(current : T) {
+        if (condition(val, current)) {
+          val := current;
+          lastWatermarkTimestamp := Time.now();
+        } else if (Time.now() > lastWatermarkTimestamp + resetIntervalSeconds * 1_000_000_000) {
+          val := current;
+        };
+      };
+    };
+
+    var count_ : Nat = 0;
+    var sum_ : Nat = 0;
+    var highWatermark_ : WatermarkTracker<Nat> = WatermarkTracker<Nat>(0, func(old, new) = new > old, watermarkResetIntervalSeconds);
+    var lowWatermark_ : WatermarkTracker<Nat> = WatermarkTracker<Nat>(0, func(old, new) = new < old, watermarkResetIntervalSeconds);
+    var lastVal_ : Nat = 0;
+
+    public func lastValue() : Nat = lastVal_;
+    public func count() : Nat = count_;
+    public func sum() : Nat = sum_;
+    public func highWatermark() : Nat = highWatermark_.value();
+    public func lowWatermark() : Nat = lowWatermark_.value();
```

Comment:
We don't have to protect our private variables from code that lives in the same file.

I think we can remove all these functions and access the underlying variables directly. Saves some lines of code and quicker to read.

Reply:
Same applies to the other value classes.

Reply:
Done